sbi/inference/snre/snre_c.py

```diff
@@ -0,0 +1,181 @@
+        discard_prior_samples: bool = False,
+        retrain_from_scratch: bool = False,
+        show_train_summary: bool = False,
+        dataloader_kwargs: Optional[Dict] = None,
+    ) -> nn.Module:
+        r"""Return classifier that approximates the ratio $p(\theta,x)/p(\theta)p(x)$.
+
+        Args:
+            K: Number of theta to classify against. Minimum 1. Similar to `num_atoms`
+                for SNRE_B except SNRE_C has an additional independently drawn sample.
+            gamma: Determines the relative weight of the sum of all $K$ dependently drawn
+                classes against the marginally drawn one. Specifically, $p(y=k) := p_K$,
+                $p(y=0) := p_0$, $p_0 = 1 - K p_K$, and finally $\gamma := K p_K / p_0$.
+            training_batch_size: Training batch size.
+            learning_rate: Learning rate for Adam optimizer.
+            validation_fraction: The fraction of data to use for validation.
+            stop_after_epochs: The number of epochs to wait for improvement on the
+                validation set before terminating training.
+            max_num_epochs: Maximum number of epochs to run. If reached, we stop
+                training even when the validation loss is still decreasing. Otherwise,
+                we train until validation loss increases (see also `stop_after_epochs`).
+            clip_max_norm: Value at which to clip the total gradient norm in order to
+                prevent exploding gradients. Use None for no clipping.
+            exclude_invalid_x: Whether to exclude simulation outputs `x=NaN` or `x=±∞`
+                during training. Expect errors, silent or explicit, when `False`.
+            resume_training: Can be used in case training time is limited, e.g. on a
+                cluster. If `True`, the split between train and validation set, the
+                optimizer, the number of epochs, and the best validation log-prob will
+                be restored from the last time `.train()` was called.
+            discard_prior_samples: Whether to discard samples simulated in round 1, i.e.
+                from the prior. Training may be sped up by ignoring such less targeted
+                samples.
+            retrain_from_scratch: Whether to retrain the conditional density
+                estimator for the posterior from scratch each round.
+            show_train_summary: Whether to print the number of epochs and validation
+                loss and leakage after the training.
+            dataloader_kwargs: Additional or updated kwargs to be passed to the training
+                and validation dataloaders (like, e.g., a collate_fn)
+
+        Returns:
+            Classifier that approximates the ratio $p(\theta,x)/p(\theta)p(x)$.
+        """
+        kwargs = del_entries(locals(), entries=("self", "__class__"))
+        kwargs["num_atoms"] = kwargs.pop("K") + 1
+        kwargs["loss_kwargs"] = {"gamma": kwargs.pop("gamma")}
```

Comment:
How about setting `self.gamma = gamma` here? Then we could remove the `loss_kwargs` passed to `_loss` and avoid changes to `snre_base.py`.

Reply:
nice idea!

Reply:
That would make the `_loss` function have hidden state for `gamma`. It doesn't have hidden state currently.

Other algorithms, like #779, will require additional hyperparameters in the loss. What do you think @janfb and @michaeldeistler? Adding `loss_kwargs` does make things more flexible for the future.

Reply:
Yes, @michaeldeistler and I discussed this and agree that it is nice to have `gamma` as kwarg in `train(...)` next to  `num_classes`. So it is good as it is now.

Reply:
Okay! done